File: Python/124.binary_tree_maximum_path_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def helper(self, root):
        if not root:
            return 0
        left = max(0, self.helper(root.left))
        right = max(0, self.helper(root.right))
        ans = root.val + max(left, right)
        bothWay = root.val + left + right
        self.res = max(self.res, bothWay, ans)
        return ans

    def maxPathSum(self, root):
        self.res = -float('inf')
        
        self.helper(root)

        return self.res

File: C++/Depth-First Search/Binary Tree Series/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int helper(TreeNode* root, int& res) {
        int maxLeft = root->left ? max(0, helper(root->left, res)) : 0;
        int maxRight = root->right ? max(0, helper(root->right, res)) : 0;
        int maxRoot = root->val + max(maxLeft, maxRight);
        res = max(res, root->val + maxLeft + maxRight);
        return maxRoot;
    }
    
    int maxPathSum(TreeNode* root) {
        int res = INT_MIN;
        helper(root, res);
        return res;
    }
};